### ISSUE TITLE
Examine rebuild on startup interferes with a manual rebuild

### ISSUE THREAD

**Issue by LennardF1989** [umbraco/Umbraco-CMS#13498]
### Which Umbraco version are you using? (Please write the *exact* version, example: 10.1.0)

All since v8, maybe even v7.

### Bug summary

Umbraco does a rebuild of indexes on startup of the website using the `RebuildOnStartupHandler.cs`, and done so for ages.

Recently, we ran into an issue where a huge Umbraco website had some weirdness on startup: after deploying to production only a portion of the content would be available on the site. A manual rebuild in the backoffice would fix that.

We were always under the impression a rebuild on startup would only occur if indexes were missing, but this comment makes it clear it will always do a full rebuild on cold-boot (which is the case after a deploy):
https://github.com/umbraco/Umbraco-CMS/blob/v11/contrib/src/Umbraco.Infrastructure/Examine/RebuildOnStartupHandler.cs#L60

While not wrong, this causes an issue if you then proceed to manually rebuild the index:
https://github.com/umbraco/Umbraco-CMS/blob/v11/contrib/src/Umbraco.New.Cms.Infrastructure/Services/IndexingRebuilderService.cs#L55

Mistake 1: If you rebuild through the UI a cache key is set, preventing you from rebuilding when a rebuild is already running. A rebuild triggered by a startup does not do this.Possible solution: Also set the same cache-key when doing the background task.

Mistake 2: Due to the above (and this is only what we can assume happens), Umbraco will fight over the DB access as well as the index files themselves, effectively leaving the index in a half-indexes state, which also leads into the final mistake.

Mistake 3: The cache key assumes an index is probably done within 5 minutes. This is a wrong assumption, as our index can take much longer than that too rebuild. So whenever we saw that the UI reported being done indexing, we would start our own indexer. But since the Examine index was secretly still hard at work, our index would only contain a portion of the actual website. Possible solution: It would make more sense to just keep resetting the cache key every time a new item is indexed, meaning that after 5 minutes of inactivity, it is assumed the index is done.

Combine all these 3 together, and you have a recipe for disaster :P

In the end, we decided to disable the `RebuildOnStartupHandler.cs` completely, since rebuilding the indexes all the time for a website this large was not needed. Now also knowing that a re-index could take longer than what the UI made us believe, we would keep track of the log to tell us if the index was done (also see: https://github.com/umbraco/Umbraco-CMS/blob/v11/contrib/src/Umbraco.New.Cms.Infrastructure/Services/IndexingRebuilderService.cs#L82). After which we would do our own indexes and all is good.

### Steps to reproduce

Not completely sure, but something like this:
- Have an Examine full rebuild on startup running on a large index: take any website, and just make it index the same content multiple times to extend it beyond the artificial 5 minutes. Don't use sleeps, the DB needs to be actively working.
- While the index is running in the background, trigger another from the Examine UI.

After 5 minutes, it will say it's done and the files written are most likely incomplete.

### Expected result / actual result

- UI will acknowledge a rebuild is already running in the background.
- UI does not make the assumption that after 5 minutes a rebuild is most likely done.
- Config to disable rebuilding on cold-boot, only create missing, without having to remove it completely with code.

**Comment by Zeegaan** [comment]
I think there are 2 issues as you list, that need to be addressed here, we should:
- Make sure the cache key is set, even when doing the background task
- Make the cache key configurable so that you can configure the time yourself

**Comment by nzdev** [comment]
An approach could be to have examine refresh the cache value every batch it completes to keep the key fresh.

**Comment by nzdev** [comment]
The database lock contention could be reduced by using some sort of change tracking or kakfa like system where every change to umbraco is an event stored in the database, then it's lock free for rebuilding examine / nucache.

Another gotcha with the current system is cold boot is linked to having processed a database instruction and having the instruction count text file present. This file should probably be one for nucache and one for examine as it's possible to miss documents in examine if nucache updates successfully but examine fails (too slow to index etc and server stops)

**Comment by LennardF1989** [comment]
@Zeegaan Agreed, although I'm also curious why the index is actually ends up in a half-indexed state. I don't have an explanation, but since we've seen it happen, it is likely it can happen again under the right circumstance.

**Comment by lukehook** [comment]
Just updating the relevant links from the original post for those looking at this. In respective order

https://github.com/umbraco/Umbraco-CMS/blob/contrib/src/Umbraco.Infrastructure/Examine/RebuildOnStartupHandler.cs#60

https://github.com/umbraco/Umbraco-CMS/blob/contrib/src/Umbraco.New.Cms.Infrastructure/Services/IndexingRebuilderService.cs#L50

https://github.com/umbraco/Umbraco-CMS/blob/contrib/src/Umbraco.New.Cms.Infrastructure/Services/IndexingRebuilderService.cs#L82

**Comment by mwillebrands** [comment]
Instead of (mis)using a cachekey to identify if a rebuild is happening, wouldn't it be better to hold a list of indexes that are currently rebuilding as a property in the ExamineIndexRebuilder? This class knows when a reindex is started and when it is finished, so it could be responsible to give information on these operations.
